### DISCUSSION TITLE
Allow `throw notFound()` in `params.parse` and `validateSearch`

### DISCUSSION THREAD

**Discussion by nstepien** [TanStack/router#2622] [Ideas]
```ts
const intRegExp = /^[1-9]\d*$/;

export const Route = createFileRoute('/posts/$postId')({
  params: {
    parse(params) {
      const { postId } = params;

      if (!intRegExp.test(postId)) {
        throw notFound();
      }

      return {
        postId: Number.parseInt(postId, 10)
      };
    },
    stringify(params) {
      return {
        postId: params.postId.toString()
      };
    }
  },
});
```

I'd love to `throw notFound()` like in the example above.

We have params we know must be valid integers above 0, UUIDs, or enum values for example.
We run assertions to ensure we can narrow the type correctly when it's passed down to `beforeLoad`, `loader`, or accessed via `Route.useParams()`.

Right now we throw an error, which means invalid params will display the error component.
Displaying the error page makes it look like there's a bug in our app, when it's not the case.
Displaying the not found page is what we'd expect, and better communicates to the end user what's happening.
Not sure why it's not allowed currently.

**Comment by caillou** [comment]
I have a similar use-case using a `_splat`, where I want to validate, that there is a maximum of 2 path segments in the `_splat`.

```ts
createRoute({
  path: 'navigation/$',
  params: {
    stringify: () => {
      // ...
    },
    parse: ({ _splat }) => {
      const [folderId, documentId, ...rest] = (_splat || '').split('/');

      if (rest.length) {
        notFound({ throw: true });
      }

      // ...

    },
  },
})
```

**Comment by lo1tuma** [comment]
I have a similar use-case. Would be nice if there would be official support for this in tanstack-router.
